os/compression/compress_read.c

```diff
@@ -379,14 +560,56 @@ int compress_init(int filfd, uint16_t offset, off_t *filelen)
 	/* Assign file length as that of uncompressed file */
 	*filelen = compression_header.binary_size;
 
+	/* Set number of blocks to use for caching */
+	if (CONFIG_DECOMPRESSION_CACHE_BLOCKS < (CUTOFF_RATIO_CACHE_BLOCKS) * (compression_header.sections - 1))
+		number_blocks_caching = CONFIG_DECOMPRESSION_CACHE_BLOCKS;
+	else
+		number_blocks_caching = ((CUTOFF_RATIO_CACHE_BLOCKS) * (compression_header.sections - 1));
+
+	/* Initialize max_accesed_count to 0 */
+	max_accessed_count = 0;
+
 #if CONFIG_COMPRESSION_TYPE == 1
-	/* Allocating memory for read and out buffer to be used for LZMA decompression */
+	/* Allocating memory for compressed_data_buffer and blockcache list to be used for LZMA decompression */
 	if (compression_header.compression_format == COMPRESSION_TYPE_LZMA) {
-		buffers.read_buffer = (unsigned char *)malloc(compression_header.blocksize + 5);
-		buffers.out_buffer = (unsigned char *)malloc(compression_header.blocksize);
+
+		compressed_data_buffer = (unsigned char *)kmm_malloc(compression_header.blocksize + 5);
+		if (!compressed_data_buffer) {
+			bmdbg("Failed kmm_malloc for read buffer\n");
+			return -ENOMEM;
+		}
+
+		blockcache = (block_cache_t *)kmm_malloc(number_blocks_caching * sizeof(block_cache_t));
+		if (!blockcache) {
+			bmdbg("Failed kmm_malloc for blockcache\n");
+			return -ENOMEM;
+		}
+
+		/* Initialize blockcache list */
+		for (int i = 0; i < number_blocks_caching; i++) {
+			blockcache[i].out_buffer = (unsigned char *)kmm_malloc(compression_header.blocksize);
+			if (!blockcache[i].out_buffer) {
+				bmdbg("Failed kmm_malloc for blockcache's out_buffer\n");
+				return -ENOMEM;
+			}
+
+			blockcache[i].block_number = -1;
+			blockcache[i].no_requests_for_block = 0;
+			blockcache[i].index_block_cache = i;
+			blockcache[i].next = &blockcache[(i + 1) % number_blocks_caching];
+			blockcache[i].prev = &blockcache[(i - 1) % number_blocks_caching];
+		}
 	}
 #endif
 
+	/* Assign head, tail and most_accessed pointers */
+	head = &blockcache[0];
+	tail = &blockcache[number_blocks_caching - 1];
+	most_accessed = tail;
+
+	head->prev = NULL;
+	tail->next = NULL;
```

Comment:
Are these not assigned earlier?

Reply:
I will remove these assignments from compress_uninit. No need for this over there.
But, in compress_init, I need to do these assignments.